source/blender/render/hydra/sceneDelegate/blenderSceneDelegate.cc

```diff
@@ -16,253 +16,299 @@ namespace blender::render::hydra {
 BlenderSceneDelegate::BlenderSceneDelegate(HdRenderIndex* parentIndex, SdfPath const& delegateID, BL::Depsgraph &b_depsgraph)
   : HdSceneDelegate(parentIndex, delegateID)
   , b_depsgraph(b_depsgraph)
-  , isPopulated(false)
+  , is_populated(false)
 {
 }
 
-std::unique_ptr<ObjectExport> BlenderSceneDelegate::objectExport(SdfPath const & id)
+void BlenderSceneDelegate::set_material(ObjectData &obj_data)
 {
-  std::string name = objects[id].name;
-  for (BL::Object &obj : b_depsgraph.objects) {
-    if (obj.name_full() == name) {
-      return std::make_unique<ObjectExport>(obj, b_depsgraph);
+  Material *material = obj_data.material();
+  if (!material) {
+    obj_data.set_material_id(SdfPath::EmptyPath());
+    return;
+  }
+  SdfPath mat_id = material_id(material);
+  if (!material_data(mat_id)) {
+    MaterialData mat_data(material);
+    GetRenderIndex().InsertSprim(HdPrimTypeTokens->material, this, mat_id);
+    mat_data.export_mtlx();
+    materials[mat_id] = mat_data;
+    LOG(INFO) << "Add material: " << mat_id << ", mtlx=" << mat_data.mtlx_path.GetResolvedPath();
+  }
+  obj_data.set_material_id(mat_id);
+}
+
+void BlenderSceneDelegate::update_material(Material *material)
+{
+  SdfPath mat_id = material_id(material);
+  MaterialData *mat_data = material_data(mat_id);
+  if (mat_data) {
+    mat_data->export_mtlx();
+    LOG(INFO) << "Update material: " << mat_id << ", mtlx=" << mat_data->mtlx_path.GetResolvedPath();
+    GetRenderIndex().GetChangeTracker().MarkSprimDirty(mat_id, HdMaterial::AllDirty);
+  }
+}
+
+void BlenderSceneDelegate::update_collection()
+{
+  HdRenderIndex &index = GetRenderIndex();
+
+  /* add new objects */
+  std::set<SdfPath> available_objects;
+  for (auto &inst : b_depsgraph.object_instances) {
+    if (inst.is_instance()) {
+      continue;
+    }
+    Object *object = (Object *)inst.object().ptr.data;
+    if (!supported_object(object)) {
+      continue;
+    }
+    SdfPath obj_id = object_id(object);
+    available_objects.insert(obj_id);
+
+    if (!is_populated) {
+      add_update_object(object, true, true, true);
+    }
+  }
+
+  if (!is_populated) {
+    return;
+  }
+
+  /* remove unused objects */
+  for (auto it = objects.begin(); it != objects.end(); ++it) {
+    if (available_objects.find(it->first) != available_objects.end()) {
+      continue;
+    }
+    LOG(INFO) << "Remove: " << it->first;
+    if (index.GetRprim(it->first)) {
+      index.RemoveRprim(it->first);
+    }
+    else {
+      index.RemoveSprim(it->second.prim_type(), it->first);
     }
+    objects.erase(it);
+    it = objects.begin();
+  }
+
+  /* remove unused materials */
+  std::set<SdfPath> available_materials;
+  for (auto &obj : objects) {
+    if (obj.second.has_data(TfToken("materialId"))) {
```

Comment:
According to docs creation of TfToken is a heavy operation. We should create every TfToken only once.